DOC: Expand usage examples in README (#177)
quantopian/trading_calendars: README.md

# trading_calendars

> ![CI](https://github.com/quantopian/trading_calendars/workflows/CI/badge.svg)

Reviewer: Let's make this link to the workflows instead of to the image?
```suggestion
[![CI](https://github.com/quantopian/trading_calendars/workflows/CI/badge.svg)](https://github.com/quantopian/trading_calendars/actions?query=workflow%3ACI)
```

(Not for this PR, but looks like no CI workflow is running on master, or I would have filtered by branch=master.)

[![PyPI version](https://img.shields.io/pypi/v/trading-calendars.svg)](https://pypi.org/project/trading-calendars/)
[![Conda version](https://img.shields.io/conda/vn/conda-forge/trading-calendars.svg)](https://anaconda.org/conda-forge/trading-calendars)

A Python library of exchange calendars, frequently used with [Zipline](https://github.com/quantopian/zipline).

## Installation

```bash
$ pip install trading-calendars
```

## Quick Start

```python
import trading_calendars as tc
import pandas as pd
import pytz
```

Get all registered calendars with `get_calendar_names`:

```python
>>> tc.get_calendar_names()[:5]
['XPHS', 'FWB', 'CFE', 'CMES', 'XSGO']
```

Get a calendar with `get_calendar`:

```python
>>> xnys = tc.get_calendar("XNYS")
```

Working with sessions:

```python
>>> xnys.is_session(pd.Timestamp("2020-01-01"))
False

> >>> xnys.next_open(pd.Timestamp("2020-01-01"))

Reviewer: How about an empty line between each pair of input/output?

Certainly subjective, but denser than my liking currently:
<img width="470" alt="image" src="https://user-images.githubusercontent.com/2123660/95791037-84fdcb00-0cae-11eb-9ea2-bae41de7407f.png">

>>> xnys.previous_close(pd.Timestamp("2020-01-01"))
Timestamp('2019-12-31 21:00:00+0000', tz='UTC')
>>> xnys.sessions_in_range(
>>>     pd.Timestamp("2020-01-01", tz=pytz.UTC),
>>>     pd.Timestamp("2020-01-10", tz=pytz.UTC)
>>> )
DatetimeIndex(['2020-01-02 00:00:00+00:00', '2020-01-03 00:00:00+00:00',
                '2020-01-06 00:00:00+00:00', '2020-01-07 00:00:00+00:00',
                '2020-01-08 00:00:00+00:00', '2020-01-09 00:00:00+00:00',
                '2020-01-10 00:00:00+00:00'],
                dtype='datetime64[ns, UTC]', freq='C')
>>> xnys.sessions_window(
>>>     pd.Timestamp("2020-01-02", tz=pytz.UTC),
>>>     7
>>> )
DatetimeIndex(['2020-01-02 00:00:00+00:00', '2020-01-03 00:00:00+00:00',
                '2020-01-06 00:00:00+00:00', '2020-01-07 00:00:00+00:00',
                '2020-01-08 00:00:00+00:00', '2020-01-09 00:00:00+00:00',
                '2020-01-10 00:00:00+00:00', '2020-01-13 00:00:00+00:00'],
                dtype='datetime64[ns, UTC]', freq='C')
```

**NOTE**: see the [TradingCalendar class](https://github.com/quantopian/trading_calendars/blob/master/trading_calendars/trading_calendar.py) for more advanced usage.

> Trading calendars also supports commandline usage, printing a unix-cal like calendar indicating which days are trading sessions or holidays.

Reviewer: ```suggestion
Trading calendars also supports command line usage, printing a unix-cal like calendar indicating which days are trading sessions or holidays.
```

# trading_calendars

![CI](https://github.com/quantopian/trading_calendars/workflows/CI/badge.svg)

> [![PyPI version](https://badge.fury.io/py/trading-calendars.svg)](https://badge.fury.io/py/trading-calendars)

Author: 🤷 It is the first choice when I googled pypi badge and is the service we use in `libpy`.  I don't have a view.

> 

Reviewer: Shall we add conda too?
```suggestion
[![Conda version](https://img.shields.io/conda/vn/conda-forge/trading-calendars.svg)](https://anaconda.org/conda-forge/trading-calendars)

```

A Python library of exchange calendars, frequently used with [Zipline](https://github.com/quantopian/zipline).

## Installation

```bash
$ pip install trading-calendars
```

## Quick Start

```python
import trading_calendars as tc
import pandas as pd
import pytz
```

Get all registered calendars with `get_calendar_names`:

```python
tc.get_calendar_names()[:5]
```
    ['XPHS', 'FWB', 'CFE', 'CMES', 'XSGO']

Instantiate a calendar with `get_calendar`:

```python
xnys = tc.get_calendar("XNYS")
```

Working with sessions:

```python
xnys.is_session(pd.Timestamp("2020-01-01"))
```

>     False

Reviewer: How about we format the results as python too?

Author: The results aren't python?

But I can put it in one big python blob with `>>>` prefixing outputs..

Reviewer: Oh, I would expect them to be python, since it's the repr of each output, which is meant to be valid python to reconstruct that object. Aren't `False` and `Timestamp('2020-01-02 14:31:00+0000', tz='UTC')` python?

> ```python
> xnys.next_open(pd.Timestamp("2020-01-01"))
> ```
> 
>     Timestamp('2020-01-02 14:31:00+0000', tz='UTC')

Reviewer: What do you think about formatting these with a prompt? (Not sure how it comes out)

```suggestion
```python
>>> xnys.next_open(pd.Timestamp("2020-01-01"))

Timestamp('2020-01-02 14:31:00+0000', tz='UTC')
```
```

xnys.previous_close(pd.Timestamp("2020-01-01"))
```

    Timestamp('2019-12-31 21:00:00+0000', tz='UTC')

```python
xnys.sessions_in_range(
    pd.Timestamp("2020-01-01", tz=pytz.UTC),
    pd.Timestamp("2020-01-10", tz=pytz.UTC)
)
```

    DatetimeIndex(['2020-01-02 00:00:00+00:00', '2020-01-03 00:00:00+00:00',
                   '2020-01-06 00:00:00+00:00', '2020-01-07 00:00:00+00:00',
                   '2020-01-08 00:00:00+00:00', '2020-01-09 00:00:00+00:00',
                   '2020-01-10 00:00:00+00:00'],
                  dtype='datetime64[ns, UTC]', freq='C')

```python
xnys.sessions_window(
    pd.Timestamp("2020-01-02", tz=pytz.UTC),
    7
)
```

    DatetimeIndex(['2020-01-02 00:00:00+00:00', '2020-01-03 00:00:00+00:00',
                   '2020-01-06 00:00:00+00:00', '2020-01-07 00:00:00+00:00',
                   '2020-01-08 00:00:00+00:00', '2020-01-09 00:00:00+00:00',
                   '2020-01-10 00:00:00+00:00', '2020-01-13 00:00:00+00:00'],
                  dtype='datetime64[ns, UTC]', freq='C')

**NOTE**: see the [TradingCalendar class](https://github.com/quantopian/trading_calendars/blob/master/trading_calendars/trading_calendar.py) for more advanced usage.

> Trading calendars also supports commandline usage, printing a unix-cal like calendar indicating which days are trading sessions.

Reviewer: ```suggestion
Trading calendars also supports commandline usage, printing a unix-cal like calendar indicating which days are trading sessions or holidays.
```

    [19] 20  21  22  23  24 [25]   [16] 17  18  19  20  21 [22]   [20] 21  22  23  24  25 [26]
    [26] 27  28  29  30  31        [23] 24  25  26  27  28 [29]   [27] 28  29  30
                                   [30] 31

            October                        November                       December
    Su  Mo  Tu  We  Th  Fr  Sa     Su  Mo  Tu  We  Th  Fr  Sa     Su  Mo  Tu  We  Th  Fr  Sa
                    1   2 [ 3]                                            1   2   3   4 [ 5]
    [ 4]  5   6   7   8   9 [10]   [ 1]  2   3   4   5   6 [ 7]   [ 6]  7   8   9  10  11 [12]
    [11] 12  13  14  15  16 [17]   [ 8]  9  10  11  12  13 [14]   [13] 14  15  16  17  18 [19]
    [18] 19  20  21  22  23 [24]   [15] 16  17  18  19  20 [21]   [20] 21  22  23  24 [25][26]
    [25] 26  27  28  29  30 [31]   [22] 23  24  25 [26] 27 [28]   [27] 28  29  30  31
                                   [29] 30

```bash
tcal XNYS 1 2020
```

            January 2020
    Su  Mo  Tu  We  Th  Fr  Sa
                [ 1]  2   3 [ 4]
    [ 5]  6   7   8   9  10 [11]
    [12] 13  14  15  16  17 [18]
    [19][20] 21  22  23  24 [25]
    [26] 27  28  29  30  31

> ## Frequenty Asked Questions

Reviewer: ```suggestion
## Frequently Asked Questions
```

### Why are open times one minute late?

> Due to its historical use in the [Zipline](https://github.com/quantopian/zipline) backtesting system, `trading_calendars` will only indicate a market is open upon the completion of the first minute bar in a day. As an example, on a regular trading day for NYSE:

Reviewer: Would it make sense to pull the zipline link out and refer to it in both places, instead of twice inline?

Author: Seems easier to have inline links rather than a footnote. I could also drop the link here.

Reviewer: Sounds fine. I don't have a strong preference on this one.

- 9:30:00 is treated as closed.
- 9:30:01 is treated as  closed.
- 9:31:00 is the first time treated as open.
- 16:00:00 is treated as open

This may change in the future.

> ## Calendar Support

Author: I purposely moved it lower. I think it is more common for people to want to look up the usage instead of looking up which calendars we support (especially since we have `get_calendar_names` now).

### Why are open times one minute late?

Due to its historical use in the [Zipline](https://github.com/quantopian/zipline) backtesting system, `trading_calendars` will only indicate a market is open upon the completion of the first minute bar in a day. As an example, on a regular trading day for NYSE:

- 9:30:00 is treated as closed.
- 9:30:01 is treated as  closed.
- 9:31:00 is the first time treated as open.

> - 16:00:00 is treated as open

Reviewer 2: Maybe add `- 16:00:01 is treated as closed`?
